### PR TITLE
[CI only]  Attempt a fix to a potential problem in RecursionOverflow 

### PR DESCRIPTION
**Pull request by AleksanderBG**
This is an attempt at a fix to the potential problem that causes #5860. The stack overflow looks as follows:
```
java.lang.StackOverflowError while compiling /tmp/1/tests/neg/i4371b.scala
java.lang.StackOverflowError
	at dotty.tools.dotc.core.RecursionOverflow.recursions(TypeErrors.scala:49)
	at dotty.tools.dotc.core.RecursionOverflow.recursions(TypeErrors.scala:49)
	at dotty.tools.dotc.core.RecursionOverflow.recursions(TypeErrors.scala:49)
	at dotty.tools.dotc.core.RecursionOverflow.recursions(TypeErrors.scala:49)
	at dotty.tools.dotc.core.RecursionOverflow.recursions(TypeErrors.scala:49)
	at dotty.tools.dotc.core.RecursionOverflow.recursions(TypeErrors.scala:49)
	at dotty.tools.dotc.core.RecursionOverflow.recursions(TypeErrors.scala:49)
	at dotty.tools.dotc.core.RecursionOverflow.recursions(TypeErrors.scala:49)
	at dotty.tools.dotc.core.RecursionOverflow.recursions(TypeErrors.scala:49)
	at dotty.tools.dotc.core.RecursionOverflow.recursions(TypeErrors.scala:49)
	at dotty.tools.dotc.core.RecursionOverflow.recursions(TypeErrors.scala:49)
...
```

I cannot replicate the problem locally in a simple way, even if I add a test like this:
```
compileFile("tests/neg/i4371b.scala", defaultOptions).times(10000)
```
Seems triggering it depends in some way on current VM state. I'll try to run the CI a few times and see if I can catch the issue.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/TypeErrors.scala [hunk 1]
```diff
@@ -40,16 +40,22 @@ class MissingType(pre: Type, name: Name) extends TypeError {
...
 
   private def recursions: List[RecursionOverflow] = {
-    val nested = previous match {
-      case previous: RecursionOverflow => previous.recursions
-      case _ => Nil
+    import scala.collection.mutable.ListBuffer
+    val result = ListBuffer.empty[RecursionOverflow]
+    @annotation.tailrec def loop(throwable: Throwable): List[RecursionOverflow] = throwable match {
+      case ro: RecursionOverflow =>
+        if (result.contains(ro)) println("caught the issue")
```
Comment (Blaisorblade) [inline compiler/src/dotty/tools/dotc/core/TypeErrors.scala]: It’d be very surprising if the `RecursionOverflow` chain was cyclic. However, the length of the chain is proportional to the stack depth, so it’s best traversed in constant stack size. To see why, you need to see how `RecursionOverflow` is created by `handleRecursive` and how that’s used in certain recursive methods in Typer (I can explain in person if desired).

Comment (AleksanderBG) [reply compiler/src/dotty/tools/dotc/core/TypeErrors.scala]: Yeah, I wasn't sure if the `RecursionOverflow` chain can actually be cyclic. If it's really proportional to stack depth, then it makes sense to construct `recursions` in a loop. How about I simply open a separate PR with that change and we merge it on its own?

Comment (Blaisorblade) [reply compiler/src/dotty/tools/dotc/core/TypeErrors.scala]: Go ahead!

**Comment by AleksanderBG** [comment]
Closing, as I could not replicate the mysterious `StackOverflow`.

**Pull request closed without merging**